### ISSUE TITLE
libp2p mesh testing

### ISSUE THREAD

**Issue by kumavis** [MetaMask/metamask-plugin#719]
In the name of facilitating libp2p-based Ethereum light client development, we're discussing running some libp2p mesh testing in the metamask extension.

Some Design Goals:
- light: minimal overhead, gentle testing
- safe: runs on a different domain than the extension (via iframe)
- big: thousands of nodes
# Progress Tracker:

- [ ] Ethereum IPLD 
  - [x] RLP
    - [x] eth-block-list
      - [x] basic
      - [x] resolve leaves (blocks)
    - [x] dump ommers
  - [ ] Simple Objects
    - [x] eth-tx
    - [ ] eth-tx-receipt
  - [ ] Eth Trie
    - [x] eth-trie rewrite
    - [x] eth-state-trie rewrite
    - [x] eth-storage-trie
    - [x] eth-tx-trie
    - [ ] eth-tx-receipt-trie
    - [x] dump transactions
    - [ ] dump transactionReceipts
- [ ] Networking
  - [x] ipfs over webrtc
  - [ ] determine the current blockchain HEAD (infura or p2p)
  - [ ] new tx, new block global pubsub
- [ ] libp2p test mesh
  - [x] add option to disable ipfs discovery
  - [ ] manually setup discovery
  - [x] update eth-ipld resolvers PR
  - [ ] transfer eth-ipld resolvers to ipld org
- [ ] Other
  - [x] blockchain - separate blockDb + detailsDb
  - [ ] block - update fromRpc
  - [ ] web3 ipfs-provider (RPC methods via IPFS lookups)
  - [ ] explore merkelization of external indexes
    - [ ] CHT (canonical hash tree)
    - [ ] index:txToBlock
    - [ ] index:logToTx
  - [ ] eth-*-trie correctly resolve keys not in trie
  - [ ] experiment with identity hash fn https://github.com/multiformats/multihash/issues/13#issuecomment-270024879

**Comment by kumavis** [comment]
node.js (non-browser) intensive test suite: https://github.com/gavinmcdermott/js-libp2p-pstn

this is a great start but has been built with a few different design goals (non-browser, short and intensive, fully connected mesh)

**Comment by kumavis** [comment]
cross-team interface/deployment will most likely happen in this way:
the extension runs a hidden iframe pointed at a domain (untrusted by the extension) controlled by the ipfs/libp2p team with some mesh test code that they can update at their will

**Comment by kumavis** [comment]
cc @diasdavid @gavinmcdermott

**Comment by kumavis** [comment]
marking as blocked until I have some concrete action items ( happy to work on some libp2p-side stuff if needed )

**Comment by kumavis** [comment]
This is blocked by this big crypto PR https://github.com/libp2p/js-libp2p-crypto/pull/12

**Comment by kumavis** [comment]
Eth-Objects

- [x] eth-block
- [ ] eth-block-list (uncles)
- [ ] eth-tx-trie
- [ ] eth-tx-receipt-trie
- [ ] eth-tx
- [ ] eth-state-trie
- [ ] eth-account-snapshot
- [ ] eth obj type heuristics research

- [ ] keccak256 codec code
  - [ ] get keccak-256 code https://github.com/multiformats/multicodec/pull/16
    - [x] make PR https://github.com/multiformats/multicodec/pull/19
  - [ ] make a PR multihash including that number
  - [ ] make a PR multihashing including that algo

- [ ] js-ipld-eth-block
  - [x] cid.toString()
  - [x] blob -> cid util.cid(ethBlock)
  - [ ] keccak256 codec code

Libp2p tests

- [ ] js-libp2p-crypto
  - [ ] blocked by https://github.com/libp2p/js-libp2p-crypto/pull/12

**Comment by kumavis** [comment]
next blockers:
- [ ] https://github.com/ipfs/js-libp2p-ipfs-browser/pull/99
- [ ] https://github.com/ipfs/js-ipfs/pull/485

**Comment by kumavis** [comment]
blocker: https://github.com/ipfs/js-libp2p-ipfs-browser/issues/122

**Comment by kumavis** [comment]
next blockers:
- [ ] js-ipld-resolver
  - [ ] add tests for eth-block

- [ ] try on js-ipfs#feat/dap-api

**Comment by kumavis** [comment]
new blocker:
- [ ] https://github.com/ipfs/js-cid/pull/7

**Comment by kumavis** [comment]
heres a new sketch of the ethereum obj types ( 8 types in total )
```
eth-block
  ommers: eth-block-list
  txs: eth-tx-trie
  receipts: eth-tx-receipt-trie
  state: eth-state-trie

eth-block-list (rlp array)
  [leaves]: eth-block

eth-tx-trie (merkle trie)
  [leaves]: eth-tx

eth-tx-receipt-trie (merkle trie)
  [leaves]: links to eth-tx-receipt

eth-state-trie (secure merkle trie)
  [leaves]: links to eth-account-snapshot

eth-tx (local data only)
eth-tx-receipt (local data only)
eth-account-snapshot (local data only)
```

**Comment by kumavis** [comment]
here is an attempt at a concise-yet-hacky set
```
[0x90] eth-block  (rlp)
  ommers: eth-rlp:eth-block
  txs: eth-trie:eth-tx
  receipts: eth-trie:eth-tx-receipt
  state: eth-trie:eth-account-snapshot

[0x91] eth-tx  (rlp) (local data only)
[0x92] eth-tx-receipt  (rlp) (local data only)
[0x93] eth-account-snapshot  (rlp) (local data only)

[0x94] eth-trie[:type] (merkle trie)
  [leaves]: buffer or <type>

[0x95] eth-rlp[:type] (rlp)
  [leaves]: buffer or <type>
```

**Comment by kumavis** [comment]
here is an ultra-concise with only 2 types, maximizing path fanciness
```
[0x90] eth-rlp[:type]
  [leaves]: buffer or <type>

[0x91] eth-trie[:type]
  [leaves]: buffer or <type>

how to lookup:

eth-block: `<cid for block as eth-rlp>/:block/`
  ommers(index): `<cid for ommersRoot as eth-rlp>/<index>/:block/`
  txs(index): `<cid for transactionsTrie as eth-trie>/<rlp.encode(index)>/:tx/`
  receipts(index): `<cid for receiptTrie as eth-trie>/<rlp.encode(index)>/:txr/`
  state(address): `<cid for stateRoot as eth-trie>/<hash(address)>/:account/`

eth-tx: `<cid for tx as eth-rlp>/:tx/`
  (local data only)

eth-tx-receipt: `<cid for txr as eth-rlp>/:txr/`
  (local data only)

eth-account-snapshot: `<cid for accountSnapshot as eth-rlp>/:account/`
  storage(key): `<cid for storageRoot as eth-trie>/<hash(key)>/`
```

**Comment by kumavis** [comment]
theres one weird thing but i realize it would be true for the 6-type and 2-type version

resolving `<cid for eth-block>/state/<address>`
would not be the same as
resolving `<cid for eth-block>/state/`
then resolving `<address>`
bc `<cid for eth-block>/state/<address>` will hash the address after the first step, resolving to  `<cid for stateRoot as eth-trie>/<hash of address>`

**Comment by kumavis** [comment]
@wanderer what do you think of the 2-type solution?

**Comment by kumavis** [comment]
zomg unified ethereum ipld proposal ( 1 type 😎  )
```
[0x90] eth-star[:type]
  {leaves}: buffer
  {leaves}/:type: <type>

how to lookup:

eth-block: `<cid for block as eth-star>/:block/`
  ommers(index): `<cid for ommersRoot as eth-star>/<index>/:block/`
  txs(index): `<cid for transactionsTrie as eth-star>/trie@<rlp.encode(index)>/:tx/`
  receipts(index): `<cid for receiptTrie as eth-star>/trie@<rlp.encode(index)>/:txr/`
  state(address): `<cid for stateRoot as eth-star>/<hash(address)>/:account/`

eth-tx: `<cid for tx as eth-star>/:tx/`
  (local data only)

eth-tx-receipt: `<cid for txr as eth-star>/:txr/`
  (local data only)

eth-account-snapshot: `<cid for accountSnapshot as eth-star>/:account/`
  storage(key): `<cid for storageRoot as eth-star>/trie@<hash(key)>/`
```

**Comment by kumavis** [comment]
that said, we have room for 16384 codes before we hit the 3 byte range, which is currently reserved for apps
so maybe we dont need to be so concise
so maybe taking up 8 types and having the prettiest paths is reasonable : )

**Comment by kumavis** [comment]
tho im sure in the future ill be interested in adding this fancy type casting for some other reason, like converting eth contract storage values into cids or other formats

**Comment by kumavis** [comment]
@wanderer btw jbenet and daviddias said dont hack the paths, better to have more formats

**Comment by kumavis** [comment]
Major TODOS:
- [ ] implement ipld resolvers for all eth objects
- [ ] be able to sync the blockchain (via rpc) and process the state changes
- [ ] dump ethereum state into IPLD objects

**Comment by kumavis** [comment]
who likes todo lists? you do!

### RLP
- [x] eth-block-list
  - [x] basic
  - [x] resolve leaves (blocks)
- [x] dump ommers

### Simple Objects
- [x] eth-tx
- [ ] eth-tx-receipt

### Eth Trie
- [x] eth-trie rewrite
- [x] eth-state-trie rewrite
- [x] eth-storage-trie
- [ ] eth-tx-trie
- [ ] eth-tx-receipt-trie
- [ ] dump transactions
- [ ] dump transactionReceipts

### Other
- [x] blockchain - separate blockDb + detailsDb
- [ ] rpc to ipfs
- [ ] eth-log ?

**Comment by kumavis** [comment]
TIL "identity hash fn" https://github.com/multiformats/multihash/issues/13#issuecomment-270024879

**Comment by kumavis** [comment]
consider using rlpx over libp2p transport

**Comment by kumavis** [comment]
proposal: blocks via ipfs-http-api on parity https://github.com/ethcore/parity/issues/4172

**Comment by kumavis** [comment]
- [ ] add option to disable ipfs discovery
 - [ ] manually setup discovery
 - [ ] update eth-ipld resolvers PR
 - [ ] transfer eth-ipld resolvers to ipld org

**Comment by kumavis** [comment]
identifying RPC method data/index requirements

```
client
  eth_protocolVersion
  eth_syncing
  eth_coinbase
  eth_mining
  eth_hashrate
  eth_gasPrice
  eth_accounts
  eth_sign
  eth_sendTransaction
  eth_sendRawTransaction
  eth_newFilter
  eth_newBlockFilter
  eth_newPendingTransactionFilter
  eth_uninstallFilter
  eth_getFilterChanges
  eth_getFilterLogs

ipfs
  eth_getBlockByHash
  eth_getUncleByBlockHashAndIndex

ipfs-walk (selectors/transforms)
  eth_getBlockTransactionCountByHash
  eth_getUncleCountByBlockHash

txToBlock
  eth_getTransactionReceipt
  eth_getTransactionByHash
  eth_getTransactionByBlockHashAndIndex

blockTracker
  eth_blockNumber

CHT
  eth_call
  eth_estimateGas
  eth_getBalance
  eth_getStorageAt
  eth_getTransactionCount
  eth_getCode
  eth_getBlockByNumber
  eth_getTransactionByBlockNumberAndIndex (+txToBlock)
  eth_getBlockTransactionCountByNumber (+ipfs-walk)
  eth_getUncleCountByBlockNumber (+ipfs-walk)
  eth_getUncleByBlockNumberAndIndex
  eth_getLogs (log-index?)


HEX String - an integer block number
String "earliest" for the earliest/genesis block
String "latest" - for the latest mined block
String "pending" - for the pending state/transactions

transaction
├─ blockHash { txToBlock }
├─ blockNumber { txToBlock }
├─ creates { txToBlock, compute, ... }
├─ from { ecrecover }
├─ gas {}
├─ gasPrice {}
├─ hash {}
├─ input {}
├─ nonce {}
├─ publicKey { ecrecover }
├─ r {}
├─ raw {}
├─ s {}
├─ to {}
├─ transactionIndex { txToBlock, blockToTxs }
├─ v {}
└─ value {}
```

**Comment by kumavis** [comment]
for those playing along at home, i made a thing: kumavis.github.io/metamask-mesh

**Comment by kumavis** [comment]
maybe relevant: https://github.com/bitcoin/bips/blob/master/bip-0122.mediawiki

**Comment by kumavis** [comment]
migrated js-ipld-eth-* modules to ipld org

**Comment by kumavis** [comment]
(a few weeks back) friends at Parity added an http server that serves parity data from CIDs. it got merged into master : )

**Comment by kumavis** [comment]
tracking parity integration here https://github.com/ipfs/js-ipfs/issues/763

**Comment by kumavis** [comment]
ipfs routing improvements https://github.com/ipfs/js-ipfs/issues/833

**Comment by kumavis** [comment]
here is the current custom rolled parity->ipfs server https://github.com/kumavis/ipfs-eth-bridge
it works if you can connect to it

**Comment by kumavis** [comment]
currently eth-ipfs bridge only serves blocks, transactions (no tries), state and storage tries, no tx receipts or no tx receipt tries. These are missing b/c parity does not index them by hash (tx receipt) or doesn't store them at all (tx trie, tx receipt tree).

the bridge is available over http as https://ipfs.lab.metamask.io

**Comment by kumavis** [comment]
what if: generic ipfs merkle tree heat map could show us what is happening on ethereum

**Comment by kumavis** [comment]
we will need to: create a special index for { tx tries, tx rxs, tx rx tries }

**Comment by hermanjunge** [comment]
what if: (in the long run): a box that users can just "_download and run_". Will be like this https://github.com/kumavis/ipfs-eth-bridge

With the idea "Contribute to the light clients mesh, making it stronger and resilient by downloading this application".

**Comment by kumavis** [comment]
merkelize and serve the Canonical Hash Table over ipfs

**Comment by hermanjunge** [comment]
_serverless_ block explorer

To make it more sexy, people gets it from IPNS.

Pros: We get more people pulling IPLD links, making them available faster.

**Comment by hermanjunge** [comment]
This thread is the sequel to this effort https://github.com/MetaMask/eth-ipfs-browser-client/issues/1

**Comment by kumavis** [comment]
closing this thread in favor of the new repo
